src/Plugin/resource/CachedResource.php

```diff
@@ -191,14 +206,24 @@ public function getControllers() {
    * {@inheritdoc}
    */
   public function index($path) {
-    return $this->subject->index($path);
+    return $this->getDataProvider()->index();
```

Comment:
If I understand correctly $this->subject was holding up until now the "main" resource. I think that it would make it more readable to have a ``$this->getPrimaryResource()->index()`` - so it's clear the CacheResource doesn't care about how the data was retrieved -- it just needs to know who's responsible for it.

Naming is hard, but few suggestions:
$this->getMatchedResource()
$this->getMainResource()
$this->getPrimaryResource() -- (my favorite)